diag_manager/fms_diag_axis_object.F90

```diff
@@ -0,0 +1,287 @@
+    TYPE(domainUG),     INTENT(in), OPTIONAL :: DomainU         !< Unstructured domain
+    CHARACTER(len=*),   INTENT(in), OPTIONAL :: aux             !< Auxiliary name, can only be <TT>geolon_t</TT>
+                                                                !! or <TT>geolat_t</TT>
+    CHARACTER(len=*),   INTENT(in), OPTIONAL :: req             !< Required field names.
+    INTEGER,            INTENT(in), OPTIONAL :: tile_count      !< Number of tiles
+    INTEGER,            INTENT(in), OPTIONAL :: domain_position !< Domain position, "NORTH" or "EAST"
+
+    obj%axis_name = diag_copy_string(name)
+    obj%units = diag_copy_string(units)
+    obj%cart_name = diag_copy_string(cart_name) !< TO DO Check for valid cart_names
+    if (present(long_name)) obj%long_name = diag_copy_string(long_name)
+
+    select type (axis_data)
+    type is (real(kind=r8_kind))
+      allocate(real(kind=r8_kind) :: obj%axis_data(size(axis_data)))
+      obj%axis_data = axis_data
+    type is (real(kind=r4_kind))
+      allocate(real(kind=r4_kind) :: obj%axis_data(size(axis_data)))
+      obj%axis_data = axis_data
+    class default
+      call mpp_error(FATAL, "The axis_data in your diag_axis_init call is not a supported type. &
+                          &  Currently only r4 and r8 data is supported.")
+    end select
+
+    !< TO DO check the presence of multiple Domains
+    if (present(Domain)) then
+      allocate(fms_domain1d_t :: obj%axis_domain)
+      call obj%axis_domain%set(Domain=Domain)
+    else if (present(Domain2)) then
+      allocate(fms_domain2d_t :: obj%axis_domain)
+      call obj%axis_domain%set(Domain2=Domain2)
+    else if (present(DomainU)) then
+      allocate(fms_domainUG_t :: obj%axis_domain)
+      call obj%axis_domain%set(DomainU=DomainU)
+    endif
+
+    obj%tile_count = 1
+    if (present(tile_count)) obj%tile_count = tile_count
+
+    !< TO DO Check for valid domain_position
+    obj%domain_position = CENTER
+    if (present(domain_position)) obj%domain_position = domain_position
+
+    obj%length = size(axis_data)
+
+    !< TO DO Check for valid direction
+    obj%direction = 0
+    if (present(direction)) obj%direction = direction
+
+    !< TO DO Check if id is valid and with the same parameters
+    obj%edges = 0
+    if (present(edges)) obj%edges = edges
+
+    if (present(aux)) obj%aux = diag_copy_string(aux)
+    if (present(req)) obj%req = diag_copy_string(req)
+
+    obj%nsubaxis = 0
+  end subroutine diag_axis_init
+
+  !> @brief Get the length of the axis
+  !> @return axis length
+  function get_axis_length(obj) &
+  result (axis_length)
+    class(diag_axis_t), intent(inout) :: obj !< diag_axis obj
+    integer                           :: axis_length
+
+    axis_length = obj%length
+
+    !< If the axis is domain decomposed axis_length will be set to the length for the current PE:
+    if (allocated(obj%axis_domain)) axis_length = obj%axis_domain%length(obj%cart_name, obj%domain_position)
```

Comment:
Should this be an if/else statement?

Reply:
`axis_length` is already defined in L199 if it is not.

Reply:
Yeah, so if `(allocated(obj%axis_domain))` you are setting `axis_length` twice instead of once.
```fortran
if (.not.allocated(obj%axis_domain)) then
    axis_length = obj%length
else
    axis_length = obj%axis_domain%length(obj%cart_name, obj%domain_position)
endif
```
Where or not it evaluates to true, the code will still check the `if` statement, so you are just adding an extra instruction.